### PR TITLE
[SofaBoundaryCondition] Factorize partial fixedconstraint

### LINKED ISSUE #440: Merge FixedConstraint and PartialFixedConstraint
Hi ! 

Many existing component (FixedPlaneConstraint, FixedRotationConstraint, PointConstraint...) are particular implementation of PartialFixedConstraint. We have put them in a separate plugin named
"Specific Boundary Conditions".

To finalize this work, it could be useful to have an implementation of FixedConstraint, on which you can define a subset of direction that you want to block. It would allow to have a unique implementation for fixed constraint in general.

What do you think ?

Christian

______________________________________________________

Suggested labels:

### PR DESCRIPTION
**Pull request by hugtalbot**
This PR factorizes the code of PartialFixedConstraint by inheriting for most functions (init, draw) from the FixedConstraint
FIX #440 

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

File: modules/SofaBoundaryCondition/PartialFixedConstraint.h [hunk 1]
```diff
@@ -73,97 +68,40 @@ class PartialFixedConstraint : public core::behavior::ProjectiveConstraintSet<Da
...
 
     // -- Constraint interface
     void init() override;
+    void reinit() override;
 
     void projectResponse(const core::MechanicalParams* mparams, DataVecDeriv& resData) override;
-    void projectVelocity(const core::MechanicalParams* mparams, DataVecDeriv& vData) override;
-    void projectPosition(const core::MechanicalParams* mparams, DataVecCoord& xData) override;
+    //void projectVelocity(const core::MechanicalParams* mparams, DataVecDeriv& vData) override;
+    //void projectPosition(const core::MechanicalParams* mparams, DataVecCoord& xData) override;
```
Comment (guparan) [inline modules/SofaBoundaryCondition/PartialFixedConstraint.h]: To be removed?

File: modules/SofaBoundaryCondition/FixedConstraint.inl [hunk 2]
```diff
@@ -252,7 +251,7 @@ void FixedConstraint<DataTypes>::projectVelocity(const core::MechanicalParams* m
     const SetIndexArray & indices = this->d_indices.getValue();
     helper::WriteAccessor<DataVecDeriv> res ( mparams, vData );
 
-    if( this->d_fixAll.getValue()==true )    // fix everyting
+    if( this->d_fixAll.getValue() )    // fix everyting
```
Comment (damienmarchal) [inline modules/SofaBoundaryCondition/FixedConstraint.inl]: why (this->) ?  few line later you use d_fixAll directly.

File: modules/SofaBoundaryCondition/FixedConstraint.inl [hunk 3]
```diff
@@ -125,43 +125,46 @@ void FixedConstraint<DataTypes>::init()
 
     topology = this->getContext()->getMeshTopology();
 
-    //  if (!topology)
-    //    serr << "Can not find the topology." << sendl;
+      if (!topology)
+        msg_warning() << "Can not find the topology";
```
Comment (damienmarchal) [inline modules/SofaBoundaryCondition/FixedConstraint.inl]: To without a topology does it crash ?

**Comment by guparan** [comment]
[ci-build][with-scene-tests]

File: modules/SofaBoundaryCondition/FixedConstraint.inl (hunk 3 above)
Comment (guparan) [reply modules/SofaBoundaryCondition/FixedConstraint.inl]: TODO: complete the message

File: modules/SofaBoundaryCondition/SofaBoundaryCondition_test/PartialFixedConstraint_test.cpp [hunk 4]
```diff
@@ -61,74 +61,94 @@ template <typename _DataTypes>
...
 
-        simulation::Node::SPtr node = createEulerSolverNode(root,"test");
+        mstate = New<sofa::component::container::MechanicalObject<DataTypes> >();
+        mstate->resize(1);
+        node->addObject(mstate);
+        createUniformMass<DataTypes>(node, *mstate.get());
 
-        typename MechanicalObject::SPtr dofs = addNew<MechanicalObject>(node);
-        dofs->resize(1);
+        std::cout<<"1"<<std::endl;
```
Comment (jnbrunet) [inline modules/SofaBoundaryCondition/SofaBoundaryCondition_test/PartialFixedConstraint_test.cpp]: Good old debug technique? ;-)

**Comment by damienmarchal** [comment]
Hello, 

On that PR I recommand that some voluntary people, not involved in the refactoring, to add some additionals tests to see if their understanding match what is implementation.

**Comment by marques-bruno** [comment]
I know this PR doesn't touch that file, but since we're looking at enhancing the tests for this component, I noticed that the way the FixedConstraint_test checks for the SparseLDLSolver (and in general how we should check for the presence of modules) could be improved

**Review by jnbrunet — APPROVED** [review]

**Pull request merged**